utils/handleToasts.ts

```diff
@@ -0,0 +1,52 @@
+import { toast } from 'react-toastify';
+
+function generateToastID() {
+  return checkIsIdOccupied(randomId());
+}
+
+function randomId() {
+  return Math.floor(Math.random() * 100000000).toString();
+}
+
+function checkIsIdOccupied(toastId: string) {
+  if (document.getElementById(toastId.toString()) !== null) {
+    return checkIsIdOccupied(randomId());
+  }
+  return toastId;
+}
+
+export function handleErrorToast(errorText: string) {
+  try {
+    const generatedToastID = generateToastID();
+    if (!document) {
+      return;
+    }
+
+    toast(errorText, {
+      type: 'error',
+      toastId: generatedToastID,
+      onOpen: () => {
+        setTimeout(() => {
+          document
+            .getElementById(generatedToastID)
+            ?.addEventListener('mouseenter', () => {
+              toast.update(generatedToastID, {
+                progress: 1,
+              });
+            });
+          document
+            .getElementById(generatedToastID)
+            ?.addEventListener('mouseleave', () => {
+              toast.update(generatedToastID, {
+                progress: 0,
+              });
+            });
+        }, 1);
+      },
```

Comment:
Tämä kohta voi olla melko naurettava :joy: 

Ilman tätä 1ms kestävää odotusta (`setTimeout`), ei tämä toiminut. Sivu ei "ehdi" lataamaan tarpeeksi nopeasti tietoja. Tämä ei voi olla kovin hyvä tapa, tuntuu vain siltä :smile: 

Kysäsin Geminiltä selitystä ja tämmöisen sain

"The 1ms timeout in your code acts as a small delay to give React time to finish rendering the toast notification component and its DOM elements. After this delay, the `document.getElementById` function has a higher chance of finding the element with the generated ID."

`onOpen`-method pitäisi tulla kutsutuksi vain silloin kun Toasti avataan käyttäjälle, joten **teoriassa** ongelmaa ei pitäisi olla `setTimeout`in kanssa, mutta en pidättelisi hengitystä :rofl: 

Tässä on `onOpen`:sta lisää: https://fkhadra.github.io/react-toastify/define-callback/

Parhain tapa lienee toteuttaa jotenkin sillä tavalla, että saadaan varma tieto onko kyseinen Toast mounted vai ei. Gemini ehdotti siihen muutamaa tapaa, mutta mikään niistä ei toiminut tässä tapauksessa :sweat_smile: 

Tuleeko jotain hyvää ideaa mieleen toteutukseen?

Reply:
Yksi varmempi muttei niin hyvä tapa on tehdä Toastille `<div>`-wrapper, jolle antaa `onMouseEnter`- ja `onMouseLeave`-propsit. 

"Ongelma" `div`-wrapperin kanssa oli se ettei Toastin aika nollautunut joka paikasta (esim. aikapalkki ja ruksi)

Reply:
Yksi vaihtoehtoinen tapa on tämmöinen:

```ts
export function handleErrorToast(errorText: string) {
  try {
    const toastId = toast(errorText, {
      type: 'error',
      onOpen: () => {
        setTimeout(() => {
          const toastEl = document.getElementById(toastId.toString());
          toastEl?.addEventListener('mouseenter', () => {
            toast.update(toastId, { progress: 1 });
          });
          toastEl?.addEventListener('mouseleave', () => {
            toast.update(toastId, { progress: 0 });
          });
        }, 1);
      },
    });
    return;
  } catch (e) {
    console.error(e);
    return e;
  }
}
```

Siinä `setTimeout` tarvitaan, jotta `toastId` muuttuja saadaan käyttöön event listenerissä.

Reply:
Tuohan oli näppärä :smile: Nyt jälkikäteen naurattaa ettei tuota tullut edes mieleen :smile: Otan tuon sun tavan käyttöön + yhdistän `window.crypto.randomUUID()`:n. Huomattavasti parempi, kiitos! :+1: 

Onko tämä 1ms:n mittainen `setTimeout` ihan ok. Lähinnä mietityttää voisiko siinä käydä seuraavasti. Sivu lataisi Toastin niin hitaasti ettei `document.getElementById()` ehdikään löytämään elementtiä?

Onneksi Toastit ladataan vasta siinä vaiheessa kun ne näytetään käyttäjälle, joten periaatteessa tuo ei pitäisi olla mahdollista